nearestneighbour/src/commonMain/kotlin/net/graphmasters/multiplatform/nearestneighbour/liveposition/LeastRecentUsedLivePositionRepository.kt

```diff
@@ -0,0 +1,132 @@
+package net.graphmasters.multiplatform.nearestneighbour.liveposition
+
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.launch
+import net.graphmasters.multiplatform.core.units.Timestamp
+
+class LeastRecentUsedLivePositionRepository(
+    private val capacity: Int = DEFAULT_CAPACITY
+): LivePositionRepository {
+
+    companion object {
+        const val REMOVE_REACHED_LIVE_POSITIONS_INTERVAL_MS: Long = 15000
+        const val DEFAULT_CAPACITY: Int = 64
+    }
+
+    private var sessions: Map<String, List<LivePosition>> = emptyMap()
+
+    private val lruSessionsOrder by lazy { LRUSessionOrder() }
+
+    private var lastReachedLivePositions: Map<String, LivePosition> = emptyMap()
+
+    private var timestampOfLastResponse: Timestamp? = null
+
+    private val coroutineScope = CoroutineScope(Dispatchers.Default)
+
+    override fun initialize() {
+        this.startRemovingReachedLivePositions()
+    }
+
+    private fun startRemovingReachedLivePositions() {
+        this.coroutineScope.launch {
+            while (true) {
+                removeReachedLivePositions()
+                delay(REMOVE_REACHED_LIVE_POSITIONS_INTERVAL_MS)
+            }
+        }
+    }
+
+    private fun removeReachedLivePositions() {
+        val filteredLastReachedLivePositions = mutableMapOf<String, LivePosition>()
+        val toRemoveSessionIDs = mutableSetOf<String>()
+
+        this.lastReachedLivePositions.forEach { (sessionID, livePosition) ->
+            if (isTimeToLiveUp(livePosition)) {
+                toRemoveSessionIDs.add(sessionID)
+            } else {
+                filteredLastReachedLivePositions[sessionID] = livePosition
+            }
+        }
+
+        this.lastReachedLivePositions = filteredLastReachedLivePositions.toMap()
+        this.removeSessions(toRemoveSessionIDs)
+    }
+
+    private fun isTimeToLiveUp(livePosition: LivePosition): Boolean =
+        this.timestampOfLastResponse?.let {
+            livePosition.timestamp + livePosition.timeToLive < it
+        } ?: false
+
+    private fun removeSessions(sessionIDs: Set<String>) {
+        sessionIDs.forEach {
+            if (this.checkIfNoMoreLivePositionsExist(it)) {
+                this.remove(it)
+            }
+        }
+    }
+
+    override fun didReachNextLivePosition(sessionID: String) {
+        this.getAndRemoveSession(sessionID)?.let {
+            this.addToReached(sessionID, it)
+        }
+    }
+
+    private fun addToReached(sessionID: String, livePos: LivePosition) {
+        val updatedMap = this.lastReachedLivePositions.toMutableMap()
+        updatedMap[sessionID] = livePos
+        this.lastReachedLivePositions = updatedMap.toMap()
```

Comment:
You don't have to convert a mutable map back to an unmutable one. `MutableMap` implements `Map`, so you can just assign or pass a `MutableMap` to a `Map`. Even prettier would be something like this:
```suggestion
        this.lastReachedLivePositions = this.lastReachedLivePositions + (sessionId to livePos)
```
Adding a map or, like in this case, a pair to a map, copies the original map

Reply:
Could be useful in a lot of cases for you i think